### PR TITLE
[#19] [Integrate] As a user, I can see my survey questions - PART 2

### LINKED ISSUE #19: [Integrate] As a user, I can see my survey questions
## Why
When a user starts taking a survey, he/she will go through a set of questions of the selected survey. The mobile app should show how many questions there are for the survey so that the user can expect how long it is going to take and he/she can always stop at any point by dismissing the screen.

## Acceptance Criteria
- Use the survey's cover image url as the background image
- Each question type has a different layout and UI which must be respected.
- Show the actual current question number and the number of questions on the Question Number label
  - `{{current-question-number}}/{{number-of-questions}}`
  - By default, the starting number is 1.
- Show the current question's text on the Question label
- Tapping the Next button should change the current question number.
  - The Question and Question Number labels should be updated in that respect
- Unit tests for Survey Question feature all passed
- At the end of the survey, submit all answers.

### LINKED ISSUE #75: [Bug] Can not log in to the App from TestFlight due to the lack of the secret file
## Issue

- Thank to Mike download the app from TestFlight to test, it can not log in to the App due to the lack of the secret file, it showed the message below:
 
IMG_1BDC4ED4D978-1

## Expected

- As a user, we can log in to the app.
 
## Steps to reproduce

1. Download app from TestFlight.
2. Open app.
3. Fill email and password.
4. Select enter and watch behavior.

### PR DESCRIPTION
**Pull request by nkhanh44**
Closed  #19 
Closed  #75

## What happened

- Add unit tests for:
  - AnswerViewModel
  - SurveyQuestionViewModel
  - CachedStorageUseCase
  - SubmissionStorageUseCase
  - CachedRepository
  - QuestionSubmissionRepository

Also created some mocks

- Configure Release Production baseConfigurationReference with Secret file.

## Insight

N/A
 
## Proof Of Work

Screenshot 2022-12-09 at 01 07 59

### CODE REVIEW

File: Survey/Sources/Supports/Extensions/Foundation/QuestionSubmission+Equatable.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//  QuestionSubmission+Equatable.swift
+//  Survey
+//
+//  Created by Khanh on 08/12/2022.
+//  Copyright © 2022 Nimble. All rights reserved.
+//
+
+import Foundation
+
+extension QuestionSubmission: Equatable {
```
Comment (blyscuit) [inline Survey/Sources/Supports/Extensions/Foundation/QuestionSubmission+Equatable.swift:11]: 👍

File: SurveyTests/Sources/Mocks/Cachable/QuestionSubmissionRepositoryMock.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    var getDataReturnValue = QuestionSubmission.dummy
+
+    var saveCalled = false
+
+    var removeDataCalled = false
+
+    var removeDataReturnValue = Observable.just(true)
+
+    var dataArgument: [QuestionSubmission]?
```
Comment (blyscuit) [inline SurveyTests/Sources/Mocks/Cachable/QuestionSubmissionRepositoryMock.swift]: This and `CachedRepositoryMock` has too much line break in the var section 😄

File: SurveyTests/Sources/Specs/Data/Repositories/CachedRepositorySpec.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            describe("its getData") {
+
+                let expectedData = APISurvey.dummyList
+
+                beforeEach {
+                    storage.performGetDataReturnValue = APISurvey.dummyList
+                    _ = repository.getData()
+                }
+
+                it("gets called") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Data/Repositories/CachedRepositorySpec.swift]: Please be more descriptive of what is called
```suggestion
                it("storage calls GetData") {
```

File: SurveyTests/Sources/Specs/Data/Repositories/CachedRepositorySpec.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            }
+
+            describe("its save") {
+
+                beforeEach {
+                    repository.save(data: APISurvey.dummyList)
+                }
+
+                it("gets called") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Data/Repositories/CachedRepositorySpec.swift]: same here

File: SurveyTests/Sources/Specs/Data/Repositories/CachedRepositorySpec.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            }
+
+            describe("its removeData") {
+
+                beforeEach {
+                    storage.performGetDataReturnValue = []
+                    _ = repository.removeData()
+                }
+
+                it("gets called") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Data/Repositories/CachedRepositorySpec.swift]: same here.

File: SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            describe("its getData") {
+
+                let expectedData = QuestionSubmission.dummy
+
+                beforeEach {
+                    storage.performGetDataReturnValue = QuestionSubmission.dummy
+                    _ = repository.getData()
+                }
+
+                it("gets called") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift]: same here

File: SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            }
+
+            describe("its save") {
+
+                beforeEach {
+                    repository.save(data: QuestionSubmission.dummy)
+                }
+
+                it("gets called") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift]: same here

File: SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            }
+
+            describe("its removeData") {
+
+                beforeEach {
+                    storage.performGetDataReturnValue = []
+                    _ = repository.removeData()
+                }
+
+                it("gets called") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift]: same here

File: SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                    clearSubmissionTrigger.send()
+                }
+
+                context("when submissionStorageUseCase calls delete") {
+
+                    it("returns output done") {
+                        expect(output.done) == ()
+                    }
+
+                    it("returns a correct response") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift]:
```suggestion
                    it("delete returns a correct response") {
```

File: SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+                    }
+                }
+
+                context("when submissionStorageUseCase calls delete") {
+
+                    it("submissionStorageUseCase calls delete") {
+                        expect(submissionStorageUseCase.deleteCalled) == true
+                    }
+
+                    it("returns correct response") {
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift]: Same

File: SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+                    it("returns output done") {
+                        expect(output.done) == ()
+                    }
+
+                    it("returns a correct response") {
+                        _ = submissionStorageUseCase.delete()
+                            .sink { _ in
+                            } receiveValue: {
+                                expect($0) == true
```
Comment (blyscuit) [inline SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift]: Somehow all of these async calls are picked up by XCTest. Normally we have to do it asynchronously or the test will be ignored, like in this [article](https://www.swiftbysundell.com/articles/unit-testing-combine-based-swift-code/). Not sure if the mock is so fast or some logic are sync underneath 🤔

**Review by blyscuit — APPROVED** [review]
Looks good. Only wording comments.

File: SurveyTests/Sources/Mocks/Cachable/QuestionSubmissionStorageMock.swift [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    var performSetDataCalled = false
+    var performGetDataReturnValue = QuestionSubmission.dummy
+    var performSetDataReturnValue = [QuestionSubmission]()
+
+    var questionsSubmission: [QuestionSubmission] {
+        get {
+            performGetDataCalled = true
+            return performGetDataReturnValue
+        }
+
```
Comment (ducbm051291) [inline SurveyTests/Sources/Mocks/Cachable/QuestionSubmissionStorageMock.swift]: This line break is unnecessary 😁

File: SurveyTests/Sources/Mocks/Cachable/SurveyListStorageMock.swift [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    var performSetDataCalled = false
+    var performGetDataReturnValue = APISurvey.dummyList
+    var performSetDataReturnValue = [APISurvey]()
+
+    var cachedSurveyList: [APISurvey] {
+        get {
+            performGetDataCalled = true
+            return performGetDataReturnValue
+        }
+
```
Comment (ducbm051291) [inline SurveyTests/Sources/Mocks/Cachable/SurveyListStorageMock.swift]: Same here

File: SurveyTests/Sources/Mocks/Submission/SubmissionStorageUseCaseMock.swift [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+import Combine
+
+@testable import Survey
+
+final class SubmissionStorageUseCaseMock: SubmissionStorageUseCaseProtocol {
+
+    var storeCalled = false
+
+    var loadCalled = false
+    var loadReturnValue = QuestionSubmission.dummy
```
Comment (ducbm051291) [inline SurveyTests/Sources/Mocks/Submission/SubmissionStorageUseCaseMock.swift:17]: Is this supposed to be Observable too? 
```suggestion
    var loadReturnValue = Observable.just(QuestionSubmission.dummy)
```

File: SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                    expect(storage.performSetDataCalled).to(beTrue())
+                }
+
+                it("returns a correct response") {
+                    _ = repository.removeData()
+                        .sink { _ in
+                        } receiveValue: {
+                            expect($0) == true
+                        }
```
Comment (ducbm051291) [inline SurveyTests/Sources/Specs/Data/Repositories/QuestionSubmissionRepositorySpec.swift]: Should we have a cancel bag here?

File: SurveyTests/Sources/Specs/Domain/Use Case/Cached/CachedStorageUseCaseSpec.swift [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+                it("repository calls removeData") {
+                    expect(repository.removeDataCalled) == true
+                }
+
+                it("repository calls removeData with correct response") {
+                    _ = repository.removeData()
+                        .sink { _ in
+                        } receiveValue: {
+                            expect($0) == true
+                        }
```
Comment (ducbm051291) [inline SurveyTests/Sources/Specs/Domain/Use Case/Cached/CachedStorageUseCaseSpec.swift]: Same here

File: SurveyTests/Sources/Specs/Domain/Use Case/Submission/SubmissionStorageUseCaseSpec.swift [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+                it("repository calls removeData") {
+                    expect(repository.removeDataCalled) == true
+                }
+
+                it("repository calls removeData with correct response") {
+                    _ = repository.removeData()
+                        .sink { _ in
+                        } receiveValue: {
+                            expect($0) == true
+                        }
```
Comment (ducbm051291) [inline SurveyTests/Sources/Specs/Domain/Use Case/Submission/SubmissionStorageUseCaseSpec.swift]: This too

File: SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift (hunk 11 above)
Comment (ducbm051291) [reply SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift]: Although we're using Combine, the test is like the traditional synchronous way. Consider applying the [article](https://www.swiftbysundell.com/articles/unit-testing-combine-based-swift-code/) @blyscuit mentioned above; I also apply this in my IC.

**Review by ducbm051291 — CHANGES_REQUESTED** [review]

File: SurveyTests/Sources/Mocks/Submission/SubmissionStorageUseCaseMock.swift (hunk 14 above)
Comment (nkhanh44) [reply SurveyTests/Sources/Mocks/Submission/SubmissionStorageUseCaseMock.swift:17]: No, it's not :D.

File: SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift (hunk 11 above)
Comment (nkhanh44) [reply SurveyTests/Sources/Specs/Presentation/Modules/SurveyQuestion/SurveyQuestionViewModelSpec.swift]: Thank you for the suggestion. It's a nice article! I fixed all the comments and applied `awaitPublisher` to Unit tests at https://github.com/nkhanh44/ic-survey-ios/pull/88/commits/97b7c269f7a799da63d092c053475bc3eee32b1b, I also updated the QuestionSubmissionStorage at https://github.com/nkhanh44/ic-survey-ios/pull/88/commits/18ba1fd38742a5d87f46e36fbae4fbfac4c60484.

**Review by ducbm051291 — APPROVED** [review]

**Pull request merged**